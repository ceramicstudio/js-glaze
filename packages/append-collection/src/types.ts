import type { TileDocument } from '@ceramicnetwork/stream-tile'
import type { Schema } from '@glazed/common'

export type CollectionContent = {
  sliceMaxItems: number
  slicesCount: number
}
export type CollectionDoc = TileDocument<CollectionContent>
export type CollectionSchema = Schema<CollectionContent>

export type SliceContent<Item> = {
  collection: string
  sliceIndex: number
  contents: Array<Item | null>
}
export type SliceDoc<Item> = TileDocument<SliceContent<Item>>
export type SliceSchema<Item> = Schema<SliceContent<Item>>
