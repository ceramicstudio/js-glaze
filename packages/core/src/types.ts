import type StreamID from '@ceramicnetwork/streamid'
import type { Definition } from '@ceramicstudio/idx-constants'

export type Aliases = Record<string, string>

export type DefinitionWithID<C extends Record<string, unknown> = Record<string, unknown>> =
  Definition<C> & { id: StreamID }

export type IndexKey = string
export type Index = Record<IndexKey, string>

export type Entry = {
  key: IndexKey
  id: string
  record: unknown
}
