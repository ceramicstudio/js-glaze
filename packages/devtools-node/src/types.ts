/**
 * Path to a local file. If relative, it will get resolved from the current working directory.
 */
export type PathInput = URL | string
