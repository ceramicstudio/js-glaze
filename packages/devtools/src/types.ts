import type { JSONSchema } from '@glazed/types'

export type ScalarSchema =
  | JSONSchema.Boolean
  | JSONSchema.Integer
  | JSONSchema.Number
  | JSONSchema.String

export type AnySchema = ScalarSchema | JSONSchema.Array | JSONSchema.Object
