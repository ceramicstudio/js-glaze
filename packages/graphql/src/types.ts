export type Doc<Content = Record<string, any>> = {
  content: Content
  id?: string
  schema?: string
}
