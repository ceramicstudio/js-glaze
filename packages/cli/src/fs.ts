import { isAbsolute, dirname, resolve } from 'node:path'
import fs from 'fs-extra'

export function resolvePath(value: string): string {
  return isAbsolute(value) ? value : resolve(process.cwd(), value)
}

export async function read<T = any>(file: string): Promise<T> {
  return (await fs.readJSON(resolvePath(file))) as T
}

export async function write(file: string, data: unknown): Promise<void> {
  const path = resolvePath(file)
  await fs.ensureDir(dirname(path))
  await fs.writeJSON(path, data)
}
