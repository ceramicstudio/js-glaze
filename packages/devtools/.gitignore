*.log
.DS_Store
.pinning.store
node_modules
coverage
dist
