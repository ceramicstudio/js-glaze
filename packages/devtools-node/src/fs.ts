import { CeramicClient } from '@ceramicnetwork/http-client'
import { Composite } from '@glazed/devtools'
import { printGraphQLSchema } from '@glazed/graphql'
import type { EncodedCompositeDefinition, RuntimeCompositeDefinition } from '@glazed/types'
import fs from 'fs-extra'
import { resolve } from 'path'
import { cwd } from 'process'
// // fs-extra is a CommonJS module
const { readJSON, writeFile, writeJSON } = fs

import type { PathInput } from './types.js'

export function getFilePath(path: PathInput): string {
  return path instanceof URL ? path.pathname : resolve(cwd(), path)
}

export async function readEncodedComposite(
  ceramic: CeramicClient | string,
  path: PathInput
): Promise<Composite> {
  const client = ceramic instanceof CeramicClient ? ceramic : new CeramicClient(ceramic)
  const file = getFilePath(path)
  const definition = (await readJSON(file)) as EncodedCompositeDefinition
  return Composite.fromJSON({ ceramic: client, definition })
}

export async function writeEncodedComposite(
  composite: Composite,
  path: PathInput
): Promise<string> {
  const file = getFilePath(path)
  await writeJSON(file, composite.toJSON())
  return file
}

export async function writeGraphQLSchema(
  definition: RuntimeCompositeDefinition,
  path: PathInput,
  readonly?: boolean
): Promise<string> {
  const file = getFilePath(path)
  await writeFile(file, printGraphQLSchema(definition, readonly))
  return file
}

export async function writeRuntimeDefinition(
  definition: RuntimeCompositeDefinition,
  path: PathInput
): Promise<string> {
  const file = getFilePath(path)

  if (file.endsWith('.json')) {
    await writeJSON(file, definition)
  } else if (file.endsWith('.js')) {
    await writeFile(
      file,
      `// This is an auto-generated file, do not edit manually
export const definition = ${JSON.stringify(definition)}`
    )
  } else if (file.endsWith('.ts')) {
    await writeFile(
      file,
      `// This is an auto-generated file, do not edit manually
import type { RuntimeCompositeDefinition } from '@glazed/types'
export const definition: RuntimeCompositeDefinition = ${JSON.stringify(definition)}`
    )
  } else {
    throw new Error('Unsupported file type: only .json, .js and .ts extensions are supported')
  }

  return file
}

export async function writeEncodedCompositeRuntime(
  ceramic: CeramicClient | string,
  definitionPath: PathInput,
  runtimePath: PathInput,
  schemaPath?: PathInput
): Promise<void> {
  const definition = await readEncodedComposite(ceramic, definitionPath)
  const runtime = definition.toRuntime()
  await writeRuntimeDefinition(runtime, runtimePath)
  if (schemaPath != null) {
    await writeGraphQLSchema(runtime, schemaPath)
  }
}

export async function composeEncodedComposites(
  ceramic: CeramicClient | string,
  source: PathInput | Array<PathInput>,
  destination: PathInput
): Promise<string> {
  const sources = Array.isArray(source) ? source : [source]
  const composites = await Promise.all(
    sources.map(async (path) => await readEncodedComposite(ceramic, path))
  )
  const file = getFilePath(destination)
  await writeEncodedComposite(Composite.compose(composites), file)
  return file
}
