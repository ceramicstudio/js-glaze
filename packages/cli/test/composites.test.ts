import { execa } from 'execa'
import { Model } from '@ceramicnetwork/stream-model'
import { CeramicClient } from '@ceramicnetwork/http-client'
import undeployedComposite from './mocks/encoded.composite.undeployed.json'
import encodedProfilesComposite from './mocks/encoded.composite.profiles.json'
import { EncodedCompositeDefinition } from '@glazed/types'
import { Composite } from '@glazed/devtools'
import fs from 'fs-extra'
import stripAnsi from 'strip-ansi'

const MODEL1_JSON =
  '{"name":"Model1","accountRelation":"list","schema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"stringPropName":{"type":"string","maxLength":80}},"additionalProperties":false,"required":["stringPropName"]}}'

const MODEL2_JSON =
  '{"name":"Model2","accountRelation":"list","schema":{"$schema":"https://json-schema.org/draft/2020-12/schema","type":"object","properties":{"stringPropName":{"type":"string","maxLength":80}},"additionalProperties":false,"required":["stringPropName"]}}'

describe('composites', () => {
  const seed = '3a6de55a5ef33d110a5a37438704b0f0cb77ca5977131775a70ffd1c23779c8c'

  describe('composite:create', () => {
    test('composite creation fails without the schemaFilePath param', async () => {
      await expect(execa('glaze', ['composite:create'])).rejects.toThrow(
        RegExp('schemaFilePath {2}A graphQL SDL definition of the Composite encoded as a')
      )
    }, 60000)

    test('composite creation fails without the did-key param', async () => {
      const create = await execa('glaze', [
        'composite:create',
        'test/mocks/composite.schema',
      ])

      expect(create.stderr.toString().includes('No controller specified')).toBe(true)
    }, 60000)

    test('composite creation succeeds', async () => {
      const create = await execa('glaze', [
        'composite:create',
        'test/mocks/composite.schema',
        `--did-key-seed=${seed}`,
      ])
      expect(create.stdout.toString().includes('"version":"1.0"')).toBe(true)
      expect(create.stdout.toString().includes('"aliases":')).toBe(true)
      expect(create.stdout.toString().includes('"views":')).toBe(true)
    }, 60000)
  })

  describe('composite:deploy', () => {
    const checkIfModelExist = async (
      ceramic: CeramicClient,
      modelStreamID: string
    ): Promise<boolean> => {
      let wasModelLoaded = false
      await Promise.race([
        new Promise((resolve) =>
          setTimeout(() => {
            resolve(true)
          }, 5000)
        ),
        Model.load(ceramic, modelStreamID).then((model) => {
          wasModelLoaded = true
          return model
        }),
      ])
      return wasModelLoaded
    }

    test('composite deployment fails without composite path param', async () => {
      const deploy = await execa('glaze', [
        'composite:deploy',
      ])
      expect(
        deploy.stderr
          .toString()
          .includes(
            `You need to pass the composite definition either in stdin or as the compositePath param`
          )
      ).toBe(true)
    }, 60000)

    test('composite deployment succeeds', async () => {
      const nonExistentModelStreamID = Object.keys(
        (undeployedComposite as EncodedCompositeDefinition).models
      )[0]
      const ceramic = new CeramicClient()
      const doesModelExist = await checkIfModelExist(ceramic, nonExistentModelStreamID)
      expect(doesModelExist).toBeFalsy()

      const deploy = await execa('glaze', [
        'composite:deploy',
        'test/mocks/encoded.composite.undeployed.json',
      ])
      expect(deploy.stderr.toString().includes(`Deploying the composite... Done!`)).toBe(true)

      expect(deploy.stdout.toString().includes(nonExistentModelStreamID))

      const doesModelExistNow = await checkIfModelExist(ceramic, nonExistentModelStreamID)
      expect(doesModelExistNow).toBeTruthy()
    }, 60000)
  })

  describe('composite:from-model', () => {
    let model1StreamID: string
    let model2StreamID: string

    beforeAll(async () => {
      const model1Create = await execa('glaze', [
        'model:create',
        MODEL1_JSON,
        `--did-key-seed=${seed}`,
      ])
      const model2Create = await execa('glaze', [
        'model:create',
        MODEL2_JSON,
        `--did-key-seed=${seed}`,
      ])
      model1StreamID = model1Create.stdout.toString().trim()
      model2StreamID = model2Create.stdout.toString().trim()
    }, 60000)

    test('composite from model fails without the list of models', async () => {
      const create = await execa('glaze', ['composite:from-model'])
      expect(create.stderr.toString().includes('Missing list of model streamIDs')).toBe(true)
    }, 60000)

    test('composite from model succeeds', async () => {
      const create = await execa('glaze', [
        'composite:from-model',
        model1StreamID,
        model2StreamID,
        `--did-key-seed=${seed}`,
      ])
      expect(create.stdout.toString().includes('"version":"1.0"')).toBe(true)
      expect(create.stdout.toString().includes('"aliases":')).toBe(true)
      expect(create.stdout.toString().includes('"views":')).toBe(true)
      expect(create.stdout.toString().includes(model1StreamID)).toBe(true)
      expect(create.stdout.toString().includes(model2StreamID)).toBe(true)
    }, 60000)
  })

  describe('composite:models', () => {
    test('composite model listing fails without composite path param', async () => {
      const models = await execa('glaze', ['composite:models'])
      expect(
        models.stderr
          .toString()
          .includes(
            'You need to pass a path to encoded composite either via an arg or through stdin'
          )
      ).toBe(true)
    }, 60000)

    test('composite model listing succeeds without formatting params', async () => {
      const models = await execa('glaze', [
        'composite:models',
        'test/mocks/encoded.composite.profiles.json',
      ])
      expect(models.stdout.toString()).toMatchSnapshot()
    }, 60000)

    test('composite model listing succeeds with --id-only param', async () => {
      const models = await execa('glaze', [
        'composite:models',
        'test/mocks/encoded.composite.profiles.json',
        '--id-only',
      ])
      expect(models.stdout.toString()).toMatchSnapshot()
    }, 60000)

    test('composite model listing succeeds with --table', async () => {
      const models = await execa('glaze', [
        'composite:models',
        'test/mocks/encoded.composite.profiles.json',
        '--table',
      ])
      expect(stripAnsi(models.stdout.toString())).toMatchSnapshot()
    }, 60000)
  })

  describe('composite:merge', () => {
    test('composite merge fails without the list of encoded composite paths', async () => {
      const create = await execa('glaze', ['composite:merge'])
      expect(create.stderr.toString().includes('Missing list of composite file paths')).toBe(true)
    }, 60000)

    test('composite merge succeeds', async () => {
      const merge = await execa('glaze', [
        'composite:merge',
        'test/mocks/encoded.composite.profiles.json',
        'test/mocks/encoded.composite.picture.post.json',
      ])
      expect(merge.stdout.toString()).toMatchSnapshot()
    }, 60000)
  })

  describe('composite:extract-model', () => {
    test('composite by extracting models fails without the composite path and at least one model param', async () => {
      const extractModelWithNoParams = await execa('glaze', [
        'composite:extract-model',
      ])
      expect(
        extractModelWithNoParams.stderr
          .toString()
          .includes('Missing composite path and at least one model to extract')
      ).toBe(true)

      const extractModelWithJustCompositePath = await execa('glaze', [
        'composite:extract-model',
        'test/mocks/encoded.composite.picture.post.json',
      ])
      expect(
        extractModelWithJustCompositePath.stderr
          .toString()
          .includes('Missing composite path and at least one model to extract')
      ).toBe(true)
    }, 60000)

    test('composite by extracting models given by streamID succeeds', async () => {
      const encodedComposite = encodedProfilesComposite as EncodedCompositeDefinition
      expect(Object.keys(encodedComposite.models).length).toEqual(3)
      const streamIDs = Object.keys(encodedComposite.models).sort()

      const extractModel = await execa('glaze', [
        'composite:extract-model',
        'test/mocks/encoded.composite.profiles.json',
        streamIDs[0],
        streamIDs[1],
      ])
      const newEncodedComposite = JSON.parse(
        extractModel.stdout.toString()
      ) as EncodedCompositeDefinition
      expect(Object.keys(newEncodedComposite.models).length).toEqual(2)
      expect(Object.keys(newEncodedComposite.models).includes(streamIDs[0])).toBeTruthy()
      expect(Object.keys(newEncodedComposite.models).includes(streamIDs[1])).toBeTruthy()
    }, 60000)

    test('composite by extracting models given by name succeeds', async () => {
      const encodedComposite = encodedProfilesComposite as EncodedCompositeDefinition
      expect(Object.keys(encodedComposite.models).length).toEqual(3)
      const ceramicClient = new CeramicClient()
      const composite = await Composite.fromJSON({
        ceramic: ceramicClient,
        definition: encodedComposite,
      })
      const modelNames = Object.values(composite.toParams().definition.models).map(
        (modelDefinition) => {
          return modelDefinition.name
        }
      )
      const extractModel = await execa('glaze', [
        'composite:extract-model',
        'test/mocks/encoded.composite.profiles.json',
        modelNames[0],
        modelNames[1],
      ])
      const newEncodedComposite = JSON.parse(
        extractModel.stdout.toString()
      ) as EncodedCompositeDefinition
      const newComposite = await Composite.fromJSON({
        ceramic: ceramicClient,
        definition: newEncodedComposite,
      })
      expect(Object.keys(newEncodedComposite.models).length).toEqual(2)
      const newModelNames = Object.values(newComposite.toParams().definition.models).map(
        (modelDefinition) => {
          return modelDefinition.name
        }
      )
      expect(newModelNames.length).toEqual(2)
      expect(newModelNames.includes(modelNames[0])).toBeTruthy()
      expect(newModelNames.includes(modelNames[1])).toBeTruthy()
    }, 60000)
  })

  describe('composite:compile', () => {
    test('composite compilation fails without the composite path and at least one output path param', async () => {
      const compileWithNoParams = await execa('glaze', ['composite:compile'])
      expect(
        compileWithNoParams.stderr.toString().includes('Missing composite path and at output path')
      ).toBe(true)

      const compileWithJustCompositePath = await execa('glaze', [
        'composite:compile',
        'test/mocks/encoded.composite.profiles.json',
      ])
      expect(
        compileWithJustCompositePath.stderr
          .toString()
          .includes('Missing composite path and at output path')
      ).toBe(true)
    }, 60000)

    test('composite compilation succeeds', async () => {
      const dirpath = 'test/test_output_files'
      const filename = 'runtime.composite.profiles'
      const compileWithJustCompositePath = await execa('glaze', [
        'composite:compile',
        'test/mocks/encoded.composite.profiles.json',
        `${dirpath}/${filename}.json`,
        `${dirpath}/${filename}.js`,
        `${dirpath}/${filename}.ts`,
      ])
      expect(
        compileWithJustCompositePath.stderr.toString().includes('Compiling the composite... Done!')
      ).toBe(true)

      const jsonRepresentation = await fs.readFile(`${dirpath}/${filename}.json`)
      const jsRepresentation = await fs.readFile(`${dirpath}/${filename}.js`)
      const tsRepresentation = await fs.readFile(`${dirpath}/${filename}.ts`)

      expect(jsonRepresentation).toMatchSnapshot()
      expect(jsRepresentation).toMatchSnapshot()
      expect(tsRepresentation).toMatchSnapshot()
    }, 60000)
  })
})
