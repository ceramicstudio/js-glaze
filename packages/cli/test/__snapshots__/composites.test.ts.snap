// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composites composite:merge composite merge succeeds 1`] = `
"- Creating the composite...
âœ” {
  \\"version\\": \\"1.0\\",
  \\"models\\": {
    \\"kjzl6hvfrbw6c87pfzkyo9luogpvxinjep6s9aer0cbz53fij3osymi9olmhsrk\\": [
      {
        \\"jws\\": {
          \\"payload\\": \\"AXESIEa5K6jYt4KzH_ULp6UYR5yaMqjmZPMnk9JR2LjQPICM\\",
          \\"signatures\\": [
            {
              \\"signature\\": \\"eBftgjlo5m4uOzOqIpTuwDaGmZl2mw4Uhs1YiFe2PLmw85KdgQCazpZNyWcJyJLQ0S50rYKU8kFdPp82lXWmBQ\\",
              \\"protected\\": \\"eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3BwaXlYVm1KbWV1TnE4WTVrMW1XSlVydDRqTVJ3cDhlVk1ndW1zS05aV3Z6I3o2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eiJ9\\"
            }
          ],
          \\"link\\": \\"bafyreicgxev2rwfxqkzr75ilu6srqr44tizkrzte6mtzhusr3c4napearq\\"
        },
        \\"linkedBlock\\": \\"omRkYXRhpGRuYW1lbkdlbmVyaWNQcm9maWxlZnNjaGVtYaVkdHlwZWZvYmplY3RlJGRlZnOibEltYWdlU291cmNlc6VkdHlwZWZvYmplY3RldGl0bGVsSW1hZ2VTb3VyY2VzaHJlcXVpcmVkgWhvcmlnaW5hbGpwcm9wZXJ0aWVzomhvcmlnaW5hbKFkJHJlZnUjLyRkZWZzL0ltYWdlTWV0YWRhdGFsYWx0ZXJuYXRpdmVzomR0eXBlZWFycmF5ZWl0ZW1zoWQkcmVmdSMvJGRlZnMvSW1hZ2VNZXRhZGF0YXRhZGRpdGlvbmFsUHJvcGVydGllc/RtSW1hZ2VNZXRhZGF0YaVkdHlwZWZvYmplY3RldGl0bGVtSW1hZ2VNZXRhZGF0YWhyZXF1aXJlZIRjc3JjaG1pbWVUeXBlZXdpZHRoZmhlaWdodGpwcm9wZXJ0aWVzpWNzcmOjZHR5cGVmc3RyaW5naW1heExlbmd0aBiWaW1pbkxlbmd0aABkc2l6ZaJkdHlwZWdpbnRlZ2VyZ21pbmltdW0BZXdpZHRoomR0eXBlZ2ludGVnZXJnbWluaW11bQFmaGVpZ2h0omR0eXBlZ2ludGVnZXJnbWluaW11bQFobWltZVR5cGWjZHR5cGVmc3RyaW5naW1heExlbmd0aBgyaW1pbkxlbmd0aAB0YWRkaXRpb25hbFByb3BlcnRpZXP0ZyRzY2hlbWF4LGh0dHBzOi8vanNvbi1zY2hlbWEub3JnL2RyYWZ0LzIwMjAtMTIvc2NoZW1hanByb3BlcnRpZXOiZG5hbWWjZHR5cGVmc3RyaW5naW1heExlbmd0aBiWaW1pbkxlbmd0aABlaW1hZ2WhZCRyZWZ0Iy8kZGVmcy9JbWFnZVNvdXJjZXN0YWRkaXRpb25hbFByb3BlcnRpZXP0a2Rlc2NyaXB0aW9ueDJBIG1vZGVsIHRvIHN0b3JlIGNvbW1vbiBwcm9maWxlLXJlbGF0ZWQgcHJvcGVydGllc29hY2NvdW50UmVsYXRpb25mc2luZ2xlZmhlYWRlcqNlbW9kZWxSzgECAXFxCwAJaG1vZGVsLXYxZnVuaXF1ZUyK7BWGwy37iVPP01drY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eg==\\"
      }
    ],
    \\"kjzl6hvfrbw6c6uyognb9t0b8a0qh3i84ga2tgn9lkeo68bmp5jtb6mo2t0i7c0\\": [
      {
        \\"jws\\": {
          \\"payload\\": \\"AXESIKHG85zLKrxfOSpcVGkZ9ck0Yxt3p0z5VLC8z-sI4Ihg\\",
          \\"signatures\\": [
            {
              \\"signature\\": \\"mRgll5v1Y68i8PUCnQzHT5xeEws4wrjbejYJYsC_8NxrYD5wIESa4eP4PmgaLg2-vm2_i1Vtkh3kv5orBmo8Cw\\",
              \\"protected\\": \\"eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3BwaXlYVm1KbWV1TnE4WTVrMW1XSlVydDRqTVJ3cDhlVk1ndW1zS05aV3Z6I3o2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eiJ9\\"
            }
          ],
          \\"link\\": \\"bafyreifby3zzzszkxrptsks4krurt5ojgrrrw55hjt4vjmf4z7vqryeima\\"
        },
        \\"linkedBlock\\": \\"omRkYXRhpGRuYW1lbVNvY2lhbFByb2ZpbGVmc2NoZW1hpWR0eXBlZm9iamVjdGUkZGVmc6JsSW1hZ2VTb3VyY2VzpWR0eXBlZm9iamVjdGV0aXRsZWxJbWFnZVNvdXJjZXNocmVxdWlyZWSBaG9yaWdpbmFsanByb3BlcnRpZXOiaG9yaWdpbmFsoWQkcmVmdSMvJGRlZnMvSW1hZ2VNZXRhZGF0YWxhbHRlcm5hdGl2ZXOiZHR5cGVlYXJyYXllaXRlbXOhZCRyZWZ1Iy8kZGVmcy9JbWFnZU1ldGFkYXRhdGFkZGl0aW9uYWxQcm9wZXJ0aWVz9G1JbWFnZU1ldGFkYXRhpWR0eXBlZm9iamVjdGV0aXRsZW1JbWFnZU1ldGFkYXRhaHJlcXVpcmVkhGNzcmNobWltZVR5cGVld2lkdGhmaGVpZ2h0anByb3BlcnRpZXOlY3NyY6NkdHlwZWZzdHJpbmdpbWF4TGVuZ3RoGJZpbWluTGVuZ3RoAGRzaXplomR0eXBlZ2ludGVnZXJnbWluaW11bQFld2lkdGiiZHR5cGVnaW50ZWdlcmdtaW5pbXVtAWZoZWlnaHSiZHR5cGVnaW50ZWdlcmdtaW5pbXVtAWhtaW1lVHlwZaNkdHlwZWZzdHJpbmdpbWF4TGVuZ3RoGDJpbWluTGVuZ3RoAHRhZGRpdGlvbmFsUHJvcGVydGllc/RnJHNjaGVtYXgsaHR0cHM6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQvMjAyMC0xMi9zY2hlbWFqcHJvcGVydGllc6RjdXJso2R0eXBlZnN0cmluZ2ltYXhMZW5ndGgY8GltaW5MZW5ndGgAZWVtb2ppo2R0eXBlZnN0cmluZ2ltYXhMZW5ndGgCaW1pbkxlbmd0aABqYmFja2dyb3VuZKFkJHJlZnQjLyRkZWZzL0ltYWdlU291cmNlc2tkZXNjcmlwdGlvbqNkdHlwZWZzdHJpbmdpbWF4TGVuZ3RoGQGkaW1pbkxlbmd0aAB0YWRkaXRpb25hbFByb3BlcnRpZXP0a2Rlc2NyaXB0aW9ueE1BIG1vZGVsIHRvIHN0b3JlIHByb3BlcnRpZXMgdGhhdCBhY2NvdW50cyB3b3VsZCBsaWtlIHRvIHNoYXJlIG9uIHNvY2lhbCBtZWRpYW9hY2NvdW50UmVsYXRpb25mc2luZ2xlZmhlYWRlcqNlbW9kZWxSzgECAXFxCwAJaG1vZGVsLXYxZnVuaXF1ZUwbEe1ikdjuaSBBeudrY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eg==\\"
      }
    ],
    \\"kjzl6hvfrbw6c9z3f540wcc55e12e6oj3rl36o4y5ieksjrm8ty0f2z647cubmt\\": [
      {
        \\"jws\\": {
          \\"payload\\": \\"AXESIG9-ZFVU7bpABugAWLtiXDDWEJNg8njHFL2OCWkhOOxJ\\",
          \\"signatures\\": [
            {
              \\"signature\\": \\"0r1TLZkLAnsk346yLakw8LyqNZCuVbJYMDso0POUijv4Vy6dVvqd6pfeB31l1pHcrgZQ2Z5CGHfGuCgtu5mXAw\\",
              \\"protected\\": \\"eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3BwaXlYVm1KbWV1TnE4WTVrMW1XSlVydDRqTVJ3cDhlVk1ndW1zS05aV3Z6I3o2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eiJ9\\"
            }
          ],
          \\"link\\": \\"bafyreidppzsfkvhnxjaan2aalc5wexbq2yijgyhspddrjpmobfuscohmje\\"
        },
        \\"linkedBlock\\": \\"omRkYXRhpGRuYW1lbVBlcnNvblByb2ZpbGVmc2NoZW1hpGR0eXBlZm9iamVjdGckc2NoZW1heCxodHRwczovL2pzb24tc2NoZW1hLm9yZy9kcmFmdC8yMDIwLTEyL3NjaGVtYWpwcm9wZXJ0aWVzpmZnZW5kZXKjZHR5cGVmc3RyaW5naW1heExlbmd0aBgqaW1pbkxlbmd0aABpYmlydGhEYXRlo2R0eXBlZnN0cmluZ2ltYXhMZW5ndGgKaW1pbkxlbmd0aABsYWZmaWxpYXRpb25zomR0eXBlZWFycmF5ZWl0ZW1zo2R0eXBlZnN0cmluZ2ltYXhMZW5ndGgYjGltaW5MZW5ndGgAbGhvbWVMb2NhdGlvbqNkdHlwZWZzdHJpbmdpbWF4TGVuZ3RoGIxpbWluTGVuZ3RoAG1uYXRpb25hbGl0aWVzpGR0eXBlZWFycmF5ZWl0ZW1zoWR0eXBlZnN0cmluZ2htYXhJdGVtcwVobWluSXRlbXMBcHJlc2lkZW5jZUNvdW50cnmjZHR5cGVmc3RyaW5naW1heExlbmd0aAJpbWluTGVuZ3RoAHRhZGRpdGlvbmFsUHJvcGVydGllc/RrZGVzY3JpcHRpb254KEEgbW9kZWwgdG8gc3RvcmUgYWNjb3VudHMnIHBlcnNvbmFsIGRhdGFvYWNjb3VudFJlbGF0aW9uZnNpbmdsZWZoZWFkZXKjZW1vZGVsUs4BAgFxcQsACWhtb2RlbC12MWZ1bmlxdWVMrJp04TX6BcXHrkrBa2NvbnRyb2xsZXJzgXg4ZGlkOmtleTp6Nk1rcHBpeVhWbUptZXVOcThZNWsxbVdKVXJ0NGpNUndwOGVWTWd1bXNLTlpXdno=\\"
      }
    ],
    \\"kjzl6hvfrbw6c89rp8jo5tzvf67g6xk5npdhagzh0xdczpcmn3vuljcds89mtyt\\": [
      {
        \\"jws\\": {
          \\"payload\\": \\"AXESICALbDx9p97pHNy041jZ7DIDk7uOfXE1rsqbQ2gR_is0\\",
          \\"signatures\\": [
            {
              \\"signature\\": \\"_MATHLeuT-e2dz4kXoJxU4BvEEitsdyf5vuUeFkp-_PnEgspXLBj2pXtkbCHZWYyVb-m29_XwJZ7_fOQidSiAg\\",
              \\"protected\\": \\"eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3BwaXlYVm1KbWV1TnE4WTVrMW1XSlVydDRqTVJ3cDhlVk1ndW1zS05aV3Z6I3o2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eiJ9\\"
            }
          ],
          \\"link\\": \\"bafyreibabnwdy7nh33urzxfu4nmnt3bsaoj3xdt5oe225su3inubd7rlgq\\"
        },
        \\"linkedBlock\\": \\"omRkYXRhpGRuYW1lZ1BpY3R1cmVmc2NoZW1hpWR0eXBlZm9iamVjdGckc2NoZW1heCxodHRwczovL2pzb24tc2NoZW1hLm9yZy9kcmFmdC8yMDIwLTEyL3NjaGVtYWhyZXF1aXJlZIFjdXJsanByb3BlcnRpZXOhY3VybKNkdHlwZWZzdHJpbmdpbWF4TGVuZ3RoGIxpbWluTGVuZ3RoAHRhZGRpdGlvbmFsUHJvcGVydGllc/RrZGVzY3JpcHRpb25vQSBQaWN0dXJlIG1vZGVsb2FjY291bnRSZWxhdGlvbmRsaXN0ZmhlYWRlcqNlbW9kZWxSzgECAXFxCwAJaG1vZGVsLXYxZnVuaXF1ZUxE8q7DfDBGRi809c5rY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eg==\\"
      }
    ],
    \\"kjzl6hvfrbw6c6zicv2kn8asuzjug6g3n90ivo1ds9jrj60kja25tefxmykzh8h\\": [
      {
        \\"jws\\": {
          \\"payload\\": \\"AXESIE1qcTBj13uTizJ8YWChIAJtSzLjTuprNH47gHYb-zJM\\",
          \\"signatures\\": [
            {
              \\"signature\\": \\"5zRQnnpPwf-6GIbEL-WA9Nts9XZjd7RMEcjJjGfejtyLWuTYvlaFr6qpsJvIWBMxnJGxmSl9OSTBS6F7U-OhCA\\",
              \\"protected\\": \\"eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3BwaXlYVm1KbWV1TnE4WTVrMW1XSlVydDRqTVJ3cDhlVk1ndW1zS05aV3Z6I3o2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eiJ9\\"
            }
          ],
          \\"link\\": \\"bafyreicnnjytay6xpojywmt4mfqkciacnvftfy2o5jvti7r3qb3bx6zsjq\\"
        },
        \\"linkedBlock\\": \\"omRkYXRhpGRuYW1lZFBvc3Rmc2NoZW1hpWR0eXBlZm9iamVjdGckc2NoZW1heCxodHRwczovL2pzb24tc2NoZW1hLm9yZy9kcmFmdC8yMDIwLTEyL3NjaGVtYWhyZXF1aXJlZIFkdGV4dGpwcm9wZXJ0aWVzoWR0ZXh0o2R0eXBlZnN0cmluZ2ltYXhMZW5ndGgZE4hpbWluTGVuZ3RoAHRhZGRpdGlvbmFsUHJvcGVydGllc/RrZGVzY3JpcHRpb25sQSBQb3N0IG1vZGVsb2FjY291bnRSZWxhdGlvbmRsaXN0ZmhlYWRlcqNlbW9kZWxSzgECAXFxCwAJaG1vZGVsLXYxZnVuaXF1ZUyr4tGBMkJPoaeAeC5rY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtwcGl5WFZtSm1ldU5xOFk1azFtV0pVcnQ0ak1Sd3A4ZVZNZ3Vtc0tOWld2eg==\\"
      }
    ]
  },
  \\"aliases\\": {},
  \\"views\\": {
    \\"account\\": {},
    \\"root\\": {},
    \\"models\\": {}
  }
}"
`;
